instant-xml-macros/src/lib.rs

```diff
@@ -54,60 +54,122 @@ impl<'a> Serializer {
         }
     }
 
-    fn add_header(&mut self, root_name: &str, output: &'a mut proc_macro2::TokenStream) {
-        output.extend(quote!(+ "<" + #root_name));
+    fn keys_set(&self) -> BTreeSet<&str> {
+        self.other_namespaces
+            .iter()
+            .map(|(k, _)| k.as_str())
+            .collect()
+    }
+
+    fn add_header(&mut self, root_name: &str, output: &'a mut TokenStream) {
+        output.extend(quote!(
+            output.push_str("<");
+            output.push_str(#root_name);
+        ));
 
         if let Some(default_namespace) = self.default_namespace.as_ref() {
-            output.extend(quote!(+ " xmlns=\"" + #default_namespace + "\""));
+            output.extend(quote!(
+                output.push_str(" xmlns=\"");
+                output.push_str(#default_namespace);
+                output.push_str("\"");
+            ));
         }
 
         let mut sorted_values: Vec<_> = self.other_namespaces.iter().collect();
         sorted_values.sort();
 
         for (key, val) in sorted_values {
-            output.extend(quote!(+ " xmlns:" + #key + "=\"" + #val + "\""));
+            output.extend(quote!(
+                output.push_str(" xmlns:");
+                output.push_str(#key);
+                output.push_str("=\"");
+                output.push_str(#val);
+                output.push_str("\"");
+            ));
         }
-
-        output.extend(quote!(+ ">"));
+        output.extend(quote!(
+            output.push_str(">");
+        ));
     }
 
-    fn add_footer(&mut self, root_name: &str, output: &'a mut proc_macro2::TokenStream) {
-        output.extend(quote!(+ "</" + #root_name + ">"));
+    fn add_footer(&mut self, root_name: &str, output: &'a mut TokenStream) {
+        output.extend(quote!(
+            output.push_str("</");
+            output.push_str(#root_name);
+            output.push_str(">");
+        ));
     }
 
     fn process_named_field(
         &mut self,
         field: &syn::Field,
-        output: &'a mut proc_macro2::TokenStream,
+        output: &'a mut TokenStream,
+        missing_prefixes: &'a mut BTreeSet<String>,
     ) {
-        let field_name = field.ident.as_ref().unwrap().to_string();
+        let mut field_name = field.ident.as_ref().unwrap().to_string();
         let field_value = field.ident.as_ref().unwrap();
+        let field_type = if let syn::Type::Path(v) = &field.ty {
+            v.path.get_ident()
+        } else {
+            todo!();
+        };
+
+        // For non-scalars ditch the field name
+        let is_scalar = Self::is_scalar(field_type.as_ref().unwrap().to_string().as_str());
```

Comment:
We should not be comparing to the type name here. Instead we should allow the type's `ToXml` implementation to decide whether it wants to use the field name. So I guess we need to pass the field name into `write_xml()` as part of the serializer state.

Reply:
done, please take a look, I rebuilt the whole thing, seems simpler now